### PR TITLE
compiler: add location information to the IR checker

### PR DESCRIPTION
**Pull request by aykevl**
This PR is extracted from #774 and will be useful to resolve a TODO in #782.

### CODE REVIEW

File: builder/build.go [hunk 1]
```diff
@@ -72,22 +69,23 @@ func Build(pkgName, outpath string, config *compileopts.Config, action func(stri
...
 	case "z":
-		err = c.Optimize(2, 2, 5) // -Oz, default
+		errs = c.Optimize(2, 2, 5) // -Oz, default
 	default:
-		err = errors.New("unknown optimization level: -opt=" + config.Options.Opt)
+		errs = []error{errors.New("unknown optimization level: -opt=" + config.Options.Opt)}
 	}
-	if err != nil {
-		return err
+	if len(errs) != 0 {
```
Comment (jaddr2line) [inline builder/build.go]:
```suggestion
	if len(errs) > 0 {
```
This would be slightly more readable as `> 0`, but this is kinda minor.

File: compiler/errors.go [hunk 2]
```diff
@@ -16,3 +20,31 @@ func (c *Compiler) makeError(pos token.Pos, msg string) types.Error {
 func (c *Compiler) addError(pos token.Pos, msg string) {
 	c.diagnostics = append(c.diagnostics, c.makeError(pos, msg))
 }
+
+// errorAt returns an error value at the location of the instruction.
+// The location information may not be complete as it depends on debug
+// information in the IR.
+func errorAt(inst llvm.Value, msg string) scanner.Error {
```
[the remark this answers is not in the document]
Comment (aykevl) [reply compiler/errors.go]: This is literally one of the built-in error type that is used in the go/* packages.

Yes I agree the error should be an `error` instead of a `string`, but that's not really compatible with the rest of the compiler at the moment. Eventually I'd like to streamline it across the compiler but that will take a while - to me, refactoring so that packages can be compiled independently is still higher priority.

File: builder/build.go (hunk 1 above)
Comment (aykevl) [reply builder/build.go]: To me it's exactly equally readable, perhaps `!= 0` is even more readable as it test for a non-zero error (`errs != nil` would also have worked but theoretically `errs != nil && len(errs) == 0` could be true). But on the other hand, I can see where you're coming from.

File: builder/error.go [hunk 3]
```diff
@@ -12,6 +12,16 @@ func (e *MultiError) Error() string {
...
 }
 
+// newMultiError returns a *MultiError if there is more than one error, or
+// returns that error directly when there is only one. Passing an empty slice
+// will lead to a panic.
+func newMultiError(errs []error) error {
+	if len(errs) > 1 {
+		return &MultiError{errs}
+	}
+	return errs[0]
```
[the remark this answers is not in the document]
Comment (aykevl) [reply builder/error.go]: Good point. Fixed.
